### PR TITLE
[Keyboard] Add Redragon K556

### PR DESCRIPTION
**Pull request by Xinayder**
First attempt at upstreaming K556 to the latest branch.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

None

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by Xinayder** [comment]
Issues so far:

- RGB doesn't work
- Keyboard matrix is shifted to the right in some rows, the 4th and 5th row keys are shifted to the right by one key, e.g., pressing S will write A, pressing numpad 6 will write 5

I have based off the K630 code already on the repo and changed to use the full size ANSI layout (104) keys.

**Comment by dexter93** [comment]
although k556 state is.. dubious in current form ( old code faking ortho + keymap being a hybrid for iso/ansi)
probably the best base for identifying the matrix coords will be [this](https://github.com/SonixQMK/qmk_firmware/blob/sn32_master/keyboards/redragon/k556/K556_via.json#L187-L194), at least as a starting point.

for rgb_matrix: K556 uses a non-standard transistor setup. Default is as follows:
```c
#define PWM_OUTPUT_ACTIVE_LEVEL PWM_OUTPUT_ACTIVE_LOW
#define RGB_OUTPUT_ACTIVE_LEVEL RGB_OUTPUT_ACTIVE_HIGH
```
you need to select the appropriate active levels for PNP/NPN transistors in the board. See [here](https://github.com/SonixQMK/qmk_firmware/blob/sn32_develop/drivers/led/sn32/rgb_matrix_sn32f24xb.c#L73-L106) for ref.

**Comment by Xinayder** [comment]
I have updated the matrix mapping on the K556. Almost everything works except the numpad DEL, which I cannot find a way to map it correctly. I mostly used the VIA coords for everything.

> you need to select the appropriate active levels for PNP/NPN transistors in the board. See [here](https://github.com/SonixQMK/qmk_firmware/blob/sn32_develop/drivers/led/sn32/rgb_matrix_sn32f24xb.c#L73-L106) for ref.

does it make a difference? I can't find it on the old code, maybe I am overlooking something.

EDIT: with regards to RGB, how can I get it to work? I know the matrix might be a bit off but no lights are turning on, only the indicator (num lock, caps lock) LEDs.

**Comment by Xinayder** [comment]
Thanks for the tip, I inverted the logic states of the transistors and RGB is working. The brightness control doesn't seem to work.

As for VIA support, I get the following error:

```
Compiling: quantum/dynamic_keymap.c                                                                quantum/dynamic_keymap.c:91:1: error: static assertion failed: "Dynamic keymaps are configured to use more EEPROM than is available."
   91 | _Static_assert((DYNAMIC_KEYMAP_EEPROM_MAX_ADDR) - (DYNAMIC_KEYMAP_MACRO_EEPROM_ADDR) >= 100, "Dynamic keymaps are configured to use more EEPROM than is available.");
      | ^~~~~~~~~~~~~~
```

If I try to use the `DYNAMIC_KEYMAP_EEPROM_MAX_ADDR` define, this is what I get:
```
Compiling: quantum/dynamic_keymap.c                                                                quantum/dynamic_keymap.c:55:13: error: expected a string after '#pragma message' [-Werror=pragmas]
   55 | #    pragma message STR(DYNAMIC_KEYMAP_EEPROM_MAX_ADDR) " > " STR((TOTAL_EEPROM_BYTE_COUNT - 1))
      |             ^~~~~~~
quantum/dynamic_keymap.c:56:6: error: #error DYNAMIC_KEYMAP_EEPROM_MAX_ADDR is configured to use more space than what is available for the selected EEPROM driver
   56 | #    error DYNAMIC_KEYMAP_EEPROM_MAX_ADDR is configured to use more space than what is available for the selected EEPROM driver
      |      ^~~~~
cc1: all warnings being treated as errors
 [ERRORS]
```

I'm using the EEPROM wear leveling feature with the `sn32_flash` driver selected.

EDIT: good news! I got the missing numpad del key to work, turns out there was an extra comma on the keymap definition file and for that reason I couldn't set the correct matrix location for the key. I also managed to get VIA to work by reducing the amount of layers to just 2 (base + FN keys). Thanks to VIA, I could find out that some of the RGB function keys were shifted to the left and it also helped to find out how to fix the missing key on the layout.

~The only problem now, I believe, is that the RGB is a bit too dim (it's on max brightness) and it flickers a bit. Some keys don't have LED lighting as well.~

EDIT2: I added `#define MATRIX_UNSELECT_DRIVE_HIGH` to `config.h` and made some modifications to the RGB matrix on `info.json`, RGB leds are bright and it seems like everything is working!

**Comment by sunbinyuan** [comment]
Are the LEDs still flickering in this new version? Waiting for this to work so I can port it to the K582.

**Comment by Xinayder** [comment]
@sunbinyuan no, they aren't. Adding `#define MATRIX_UNSELECT_DRIVE_HIGH` to `config.h` fixed the problem.

I believe that the port is roughly complete. VIA works with 2 layers only, but that's not a big deal. RGB is working on every key. I tested it with the ANSI version, I believe the ISO version would require some changes.

Feel free to try to adapt the changes to your K582!

**Comment by sunbinyuan** [comment]
Awesome to know, thanks! Curiously, is 2 layers the maximum number supported by the eprom?

**Comment by Xinayder** [comment]
It was the maximum I could get it to build. It's weird, the K630 uses the same driver and chip and has more layers. As mentioned on the comments above, I was having EEPROM size issues when using 4 or 5 layers.

File: keyboards/redragon/k556/info.json [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+            "cycle_spiral" : true,
+            "digital_rain" : true,
+            "gradient_left_right" : true,
+            "gradient_up_down" : true,
+            "raindrops" : true,
+            "solid_reactive_simple" : true,
+            "splash" : true,
+            "typing_heatmap" : true
+        },
+        "driver": "SN32F24xB",
```
Comment (dexter93) [inline keyboards/redragon/k556/info.json]:
```suggestion
        "driver": "sn32f24xb",
```
see https://github.com/SonixQMK/qmk_firmware/commit/d62cda8c08685b902d9e3b02276398a8bc61d56c

File: keyboards/redragon/k556/keymaps/via/config.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+/* Enable VIA RGB matrix and configure EEPROM emulation for VIA */
+#define VIA_QMK_RGBLIGHT_ENABLE
```
Comment (dexter93) [inline keyboards/redragon/k556/keymaps/via/config.h]:
```suggestion
```
not needed anymore

File: keyboards/redragon/k556/config.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+#pragma once
+
+#define MATRIX_UNSELECT_DRIVE_HIGH
+
+/* key matrix size */
+#define RGB_MATRIX_LED_COUNT 106
+
+#define RGB_MATRIX_ROW_PINS { C0, C1, C3, C4, C5, C6, C7, C8, C9, C10, C11, C12, B6, B7, B8, B9, B10, B11 }
```
Comment (dexter93) [inline keyboards/redragon/k556/config.h]:
```suggestion
#define SN32_RGB_MATRIX_ROW_PINS { C0, C1, C3, C4, C5, C6, C7, C8, C9, C10, C11, C12, B6, B7, B8, B9, B10, B11 }
```

File: keyboards/redragon/k556/config.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+/* Enable additional RGB effects                                       */
+#define RGB_MATRIX_FRAMEBUFFER_EFFECTS
+#define RGB_MATRIX_KEYPRESSES
+
+/* Configure the effects:                                              */
+#define RGB_MATRIX_TYPING_HEATMAP_DECREASE_DELAY_MS 50
+
+/* Configure transistor logic for RGB matrix */
+#define PWM_OUTPUT_ACTIVE_LEVEL PWM_OUTPUT_ACTIVE_HIGH
+#define RGB_OUTPUT_ACTIVE_LEVEL RGB_OUTPUT_ACTIVE_LOW
```
Comment (dexter93) [inline keyboards/redragon/k556/config.h]:
```suggestion
#define SN32_PWM_OUTPUT_ACTIVE_LEVEL SN32_PWM_OUTPUT_ACTIVE_HIGH
#define SN32_RGB_OUTPUT_ACTIVE_LEVEL SN32_RGB_OUTPUT_ACTIVE_LOW
```

File: keyboards/redragon/k556/info.json [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    "url": "https://www.redragonzone.com/products/redragonk556",
+    "maintainer": "Xinayder",
+    "bootloader": "sn32-dfu",
+    
+    "processor": "SN32F248BF",
+    "usb": {
+        "vid": "0x0C45",
+        "pid": "0x5004",
+        "device_version": "1.0.0",
+        "force_nkro": true
```
Comment (dexter93) [inline keyboards/redragon/k556/info.json]:
```suggestion
        "device_version": "1.0.0"
```
is there a reason nkro is enforced by default?

**Review by dexter93 — CHANGES_REQUESTED** [review]
code reviewed upon latest codebase

File: keyboards/redragon/k556/info.json (hunk 5 above)
Comment (Xinayder) [reply keyboards/redragon/k556/info.json]: I just based off the previous code, which had NKRO enabled. After a quick search, the keyboard does support NKRO. Is it a problem if it's enabled by default?

Comment (dexter93) [reply keyboards/redragon/k556/info.json]: NKRO is supported and enabled, but managed by the host. Enforcing it on by default might introduce issues with dumb hosts like BIOS and apparently even macOS.

**Comment by Xinayder** [comment]
I've applied the suggestions and it's compiling and working.

**Comment by bladyle** [comment]
Can I compile and flash this to K556 with SN32F248 MCU?

**Comment by Xinayder** [comment]
> Can I compile and flash this to K556 with SN32F248 MCU?

yes.

**Comment by sunbinyuan** [comment]
> Are the LEDs still flickering in this new version? Waiting for this to work so I can port it to the K582.

I finally got around and tried porting this to k582. After comparing with the old branch `sn32_master_stable`, I noticed there weren't a lot of changes, if any. I proceeded to build the firmware for the k582 then flashed it. However, it seem the firmware wasn't very stable, and as I started typing faster, the keyboard started freezing (with spasming RGB light) and wouldn't accept any input unless I unplug and replug the keyboard. I also tried with the unchanged k556 firmware from `sn32_develop` but I am getting the same issue.

I tried going back to an older build from the k582 from the `sn32_master_stable` branch and don't seem to have any issue with keyboard freezing, except for the flickering. Really not sure what is happening.

Is there anyway to debug the reason of the freeze?

My branch here (merged the latest QMK changes):  https://github.com/sunbinyuan/sonixqmk_firmware/tree/sn32_develop

**Comment by gwagner613** [comment]
Would this work for k556 dharma, the tkl wired version?

**Comment by Xinayder** [comment]
> Would this work for k556 dharma, the tkl wired version?

If the hardware mapping didn't change, then yes (that is, if the RGB/key matrix didn't change and use the same mappings as the 100% version). You can try uploading the firmware file to your keyboard and check if everything works.

**Pull request merged**